helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java

```diff
@@ -780,13 +780,10 @@ public void disconnect() {
 
       // TODO reset user defined handlers only
       // TODO Fix the issue that when connection disconnected, reset handlers will be blocked. -- JJ
-      // This is because reset logic contains ZK operations.
-      resetHandlers(true);
-
-      if (_leaderElectionHandler != null) {
-        _leaderElectionHandler.reset(true);
+      // remove every subscribed listener
+      if (_zkclient != null) {
+        _zkclient.unsubscribeAll();
```

Comment:
Have you make sure the listeners will be add back once handleNewSession?

Reply:
Good point! I will add the test (expire the session and compare the callhandlers before and after)

Reply:
#681 is checked and it will be used to test details of this change

Reply:
Added the test case of validating callback handlers before and after session expiry in `TestControllerLeaderhshipChange`